src/components/MultiHopTrade/components/TradeConfirm/helpers.ts

```diff
@@ -0,0 +1,147 @@
+  switch (hopExecutionState) {
+    case HopExecutionState.Pending:
+      return null // No summary step for pending state
+    case HopExecutionState.AwaitingAllowanceReset:
+      return 'trade.awaitingAllowanceReset'
+    case HopExecutionState.AwaitingAllowanceApproval:
+      return 'trade.awaitingApproval'
+    case HopExecutionState.AwaitingPermit2:
+      return 'trade.awaitingPermit2Approval'
+    case HopExecutionState.AwaitingSwap:
+      return ['trade.awaitingSwap', { swapperName }]
+    case HopExecutionState.Complete:
+      return 'trade.complete'
+    default:
+      assertUnreachable(hopExecutionState)
+  }
+}
+
+type TradeStepParams = {
+  firstHopAllowanceApproval: ApprovalExecutionMetadata
+  firstHopPermit2: Omit<ApprovalExecutionMetadata, 'txHash' | 'isInitiallyRequired'> & {
+    permit2Signature?: string | undefined
+  }
+  firstHopAllowanceReset: ApprovalExecutionMetadata
+  lastHopAllowanceApproval: ApprovalExecutionMetadata
+  lastHopPermit2: Omit<ApprovalExecutionMetadata, 'txHash' | 'isInitiallyRequired'> & {
+    permit2Signature?: string | undefined
+  }
+  lastHopAllowanceReset: ApprovalExecutionMetadata
+  isMultiHopTrade?: boolean
+}
+
+export enum TradeStep {
+  FirstHopReset = 'firstHopReset',
+  FirstHopApproval = 'firstHopApproval',
+  FirstHopSwap = 'firstHopSwap',
+  LastHopReset = 'lastHopReset',
+  LastHopApproval = 'lastHopApproval',
+  LastHopSwap = 'lastHopSwap',
+  TradeComplete = 'tradeComplete',
+}
+
+export const getTradeSteps = (params: TradeStepParams): Record<TradeStep, boolean> => {
+  const {
+    firstHopAllowanceReset,
+    firstHopAllowanceApproval,
+    firstHopPermit2,
+    lastHopAllowanceReset,
+    lastHopAllowanceApproval,
+    lastHopPermit2,
+    isMultiHopTrade,
+  } = params
+  return {
+    [TradeStep.FirstHopReset]:
+      firstHopAllowanceReset.isRequired === true || firstHopAllowanceReset.txHash !== undefined,
+    [TradeStep.FirstHopApproval]:
+      firstHopAllowanceApproval.isInitiallyRequired === true ||
+      firstHopPermit2.isRequired === true ||
+      firstHopAllowanceApproval.txHash !== undefined ||
+      firstHopPermit2.permit2Signature !== undefined,
+    [TradeStep.FirstHopSwap]: true,
+    [TradeStep.LastHopReset]:
+      isMultiHopTrade === true &&
+      (lastHopAllowanceReset.isRequired === true || lastHopAllowanceReset.txHash !== undefined),
+    [TradeStep.LastHopApproval]:
+      isMultiHopTrade === true &&
+      (lastHopAllowanceApproval.isInitiallyRequired === true ||
+        lastHopPermit2.isRequired === true ||
+        lastHopAllowanceApproval.txHash !== undefined ||
+        lastHopPermit2.permit2Signature !== undefined),
+    [TradeStep.LastHopSwap]: isMultiHopTrade === true,
+    [TradeStep.TradeComplete]: true,
+  }
+}
+
+export const countTradeSteps = (params: TradeStepParams): number => {
+  return Object.values(getTradeSteps(params)).filter(Boolean).length
+}
+
+const isInApprovalState = (state: HopExecutionState): boolean => {
```

Comment:
same comment here re: permit2 step may not require an approval, it may be *only* a permit2 EIP-712 signing

Same thing, unrelated to this diff